rasa/core/brokers/kafka.py

```diff
@@ -210,15 +216,22 @@ def _publish(self, event: Dict[Text, Any]) -> None:
         )
 
         if self.producer is not None:
-            self.producer.send(
+            self.producer.produce(
                 self.topic, value=event, key=partition_key, headers=headers
             )
 
-    def _close(self) -> None:
-        if self.producer is not None:
-            self.producer.close()
-
     @rasa.shared.utils.common.lazy_property
     def rasa_environment(self) -> Optional[Text]:
         """Get value of the `RASA_ENVIRONMENT` environment variable."""
         return os.environ.get("RASA_ENVIRONMENT", "RASA_ENVIRONMENT_NOT_SET")
+
+
+def kafka_error_callback(err: "KafkaError") -> None:
+    """Callback for Kafka errors.
+
+    Any exception raised from this callback will be re-raised from the
+    triggering flush() call.
+    """
+    from confluent_kafka import KafkaException
+
+    raise KafkaException(err)
```

Comment:
I don't think we should raise on all errors. raising on the error will mean that the publishing will be stopped (?)

the docs mention that the errors should be "considered informational since the client will automatically try to recover" https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#serializingproducer

Reply:
we could add an `on_delivery` callable to calls to `produce` to handle the case where a message permanently fails delivery

Reply:
> I don't think we should raise on all errors.

Addressed: https://github.com/RasaHQ/rasa/pull/11765/commits/ad960d46c115ca8433779f2ab8e539d2449cac92
I wasn't sure what should `on_delivery` callable do? Would it be ok just to log an error with the details of the message?

Reply:
Added `on_delivery`: https://github.com/RasaHQ/rasa/pull/11765/commits/e8f9aa8f489e02d4f14a337160d70405fa118321